antipackage.py

```diff
@@ -1,130 +1,401 @@
+        if self.data:
+            src = self.data['commit']['sha']
+        else:
+            pin('/'.join(self.index[1:]))
+            src = 'master'
+        path.append(src)
+        return 'https://api.github.com/'+'/'.join(path)
+
+class GitHubRepo(object):
+
+    host = {'top': 'https://api.github.com/repos/{1}/{2}/',
+            'branch': 'branches/{0}',
+            'sha': 'commits/{0}',
+            'tag': 'tags'}
+
+    def __init__(self, pathlist, name='branch', value='master'):
+        """Stores key GitHub repository information
+
+        Parameters
+        ----------
+        pathlist : list
+            Of the form `[github, username, repository]`
+            and indicates the repository that should be
+            accessed
+        name : str
+            Should be either 'sha', 'tag' or 'branch' and
+            indicates the version access method for the repo
+        value : str
+            specific to `name` and indicated the specific
+            repository version being referenced"""
+        self.name = name
+        self.value = value
+        self.url = self._url(pathlist)
+        self._git_init()
+
+    def _url(self, pathlist):
+        """Format the GitHub url"""
+        url = self.host['top'].format(*pathlist)
+        url += self.host[self.name].format(self.value)
+        return url
+
+    def _git_init(self):
+        """Initialize data from GitHub"""
+        temp_file = self.fetch()
+        with open(temp_file, 'r') as f:
+            data = json.load(f)
+            self.sha = self._sha(data)
+
+    def fetch(self):
+        """Retrieve data from GitHub"""
+        temp_file, response = urlretrieve(self.url)
+        if not response['Status'] == '200 OK':
+            e = 'at '+self.url
             try:
-                self._install_module(fullname)
-            except InstallError:
-                if os.path.isfile(self._install_path(fullname)):
-                    print('Using existing version: ', fullname)
-                else:
-                    print('Error installing/updating module: ', fullname)
+                with open(temp_file, 'r') as f:
+                    out = json.load(f)
+                e = "message "+e+" = '"+out['message']+"'"
+            except:
+                pass
+            raise GitHubError(e)
+        return temp_file
+
+    def _sha(self, data):
+        """Return the sha"""
+        if self.name!='sha':
+            if self.name == 'tag':
+                data = self._handle_tag(data)
+            return data['commit']['sha']
+        else:
+            return self.value
+
+    def _handle_tag(self, data):
+        """Retrieve data from a """
+        for tag in data:
+            if tag['name']==self.value:
+                return tag
+        raise RetrieveError("No tag with name: '%s'" % value)
+
+# - - - - - - - - - - - -
+# Administrative Importer
+# - - - - - - - - - - - -
+
+class Importer(object):
+
+    github = GitHubHook
```

Comment:
I would put this stuff into a dictionary `hooks = {'github': GitHubHook}` and load it from there below.